config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,113 @@ type BareMetalPlatformStatus struct {
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Peers []OpenStackAPIBGPPeer `json:"peers"`
+}
+
+type OpenStackAPIBGPConfiguration struct {
+	// apiVIPs is a list of IP addresses which will be published via BGP for
+	// use by the API server. It may contain both IPv4 and IPv6 addresses.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	APIVIPs []string `json:"apiVIPs"`
+
+	// speakers is a list of BGP speaker configurations. We require a
+	// speaker configuration for every failure domain hosting a control
+	// plane node.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=failureDomain
+	Speakers []OpenStackAPIBGPSpeaker `json:"speakers"`
+}
+
+// OpenStackFailureDomain specifies a failure domain for an OpenStack server.
+// Specifically, it specifies a set of values which will be set on all machines
+// using the failure domain.
+type OpenStackFailureDomain struct {
+	// name is the unique name of this failure domain.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// computeZone specifies the OpenStack Compute availability zone for
+	// all servers in this failure domain.
+	//
+	// If not specified, the servers are provisioned in the default Compute
+	// availabity zone.
+	// +kubebuilder:validation:Optional
+	ComputeZone string `json:"computeZone,omitempty"`
+
+	// storageZone specifies the OpenStack storage availability zone for
+	// all volumes in this failure domain.
+	//
+	// If not specified, the volumes are provisioned in the default storage
+	// availabity zone.
+	// +kubebuilder:validation:Optional
+	StorageZone string `json:"storageZone,omitempty"`
+
+	// subnetID specifies an OpenStack subnet ID which will be attached as
+	// the first NIC of every server in this failure domain.
+	//
+	// +kubebuilder:validation:Optional
+	SubnetID string `json:"subnetID,omitempty"`
+}
+
 // OpenStackPlatformSpec holds the desired state of the OpenStack infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type OpenStackPlatformSpec struct{}
+type OpenStackPlatformSpec struct {
+	// failureDomains is a list of failure domains available to Machines.
+	// Values specified in a failure domain will be set on Machines in that
+	// failure domain.
+	//
+	// +listType=map
+	// +listMapKey=name
+	FailureDomains []OpenStackFailureDomain `json:"failureDomains,omitempty"`
+
+	// apiLoadBalancerType specifies the type of loadbalancer which will be
+	// configured for the API server. Permitted values are `vrrp`, `bgp`,
+	// and no value. If no value is set, `vrrp` will be assumed.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum:=vrrp;bgp
+	// +kubebuilder:validation:default:=vrrp
+	APILoadBalancerType string `json:"apiLoadBalancerType,omitempty"`
+
+	// bgpConfiguration describes the configuration of a BGP load balancer
+	// for the API server. It is only used if apiLoadBalancer is set to
```

Comment:
This suggests you need a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions).

You should create a structure like:

```
spec:
  failureDomains: []
  apiLoadBalancer:
    type: VRRP | BGP
    bgp:
      ...
```

Reply:
what about removing `apiLoadBalancerType`, and imply the use of BGP when the configuration structure `bgp` (previously known as `bgpConfiguration`) is present?

Reply:
> imply

Kube APIs are declarative, we don't imply anything!

What would you do in that case if you added a `vrrp` struct sometime in the future?

Reply:
> Kube APIs are declarative, we don't imply anything!

We have to strike a balance between being simple-but-implicit (i.e. "you added the config for BGP, so you probably want BGP") and being explicit-but-redundant (redundancy in APIs leads to incompatible or confusing combinations: what if you have `type: vrrp` _and_ you set the `bgp` config field?).

> What would you do in that case if you added a vrrp struct sometime in the future?

In my proposed solution, a potential future `vrrp` config field would be incompatible (in validation) with a `bgp` config structure. Just like in your proposed solution, a `bgp` config struct is either incompatible or confusingly cohabitant with a setting of `type: vrrp`.

I honestly don't know if my proposed solution is "overall better" than yours; if you are more confident in yours than I am in mine, then I am happy to comply.

Reply:
> redundancy in APIs leads to incompatible or confusing combinations: what if you have type: vrrp and you set the bgp config field

This is why we use [discriminated unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) in kube, there's a whole kep describing how they behave to avoid this kind of confusion.

The idea is that you use the discriminant to determine which of the fields will be set, and kube will validate that it and only it is set.

So if you had `type: vrrp` and `bgp` configured the API would error and not allow you to persist that configuration.

This is a common pattern we are using in Kube APIs so "more confident", possibly? Lets stick to the pattern. Please read through the conventions I've linked which explain how to use them

Reply:
done